app/src/main/java/ch/hikemate/app/model/route/HikesViewModel.kt

```diff
@@ -0,0 +1,1166 @@
+      viewModelScope.launch { unsaveHikeAsync(hikeId, onSuccess, onFailure) }
+
+  /**
+   * Sets the planned date of a hike.
+   *
+   * If the hike is not saved yet and the provided date is not null (i.e. the user wants to save the
+   * hike), the hike will be saved with the provided date.
+   *
+   * If the hike is already saved, the hike will be updated with the new date.
+   *
+   * Setting a null date on an unsaved hike will have no effect.
+   *
+   * @param hikeId The ID of the hike to set the planned date of.
+   * @param date The planned date for the hike.
+   * @param onSuccess To be called if the hike's date is successfully updated.
+   * @param onFailure To be called if a problem prevents the operation's success.
+   */
+  fun setPlannedDate(
+      hikeId: String,
+      date: Timestamp?,
+      onSuccess: () -> Unit = {},
+      onFailure: () -> Unit = {}
+  ) = viewModelScope.launch { setPlannedDateAsync(hikeId, date, onSuccess, onFailure) }
+
+  /**
+   * Replaces the current hikes in [hikeFlows] with hikes loaded from OSM in the provided bounds.
+   *
+   * Calling this function only launches a request asynchronously, the function will return before
+   * the hikes have actually been updated. Observe [hikeFlows] to obtain the changed list.
+   *
+   * @param bounds The minimum and maximum latitude/longitude of the rectangle to search in.
+   * @param onSuccess To be called when hikes have been successfully loaded.
+   * @param onFailure Will be called if an error is encountered.
+   */
+  fun loadHikesInBounds(
+      bounds: BoundingBox,
+      onSuccess: () -> Unit = {},
+      onFailure: () -> Unit = {}
+  ) =
+      viewModelScope.launch {
+        // Let the user know a heavy load operation is being performed
+        _loading.value = true
+
+        loadHikesInBoundsAsync(bounds, onSuccess, onFailure)
+
+        // The heavy loading operation is done now
+        _loading.value = false
+      }
+
+  /**
+   * Retrieves the bounding box and way points of the currently loaded hikes.
+   *
+   * Filters the [hikeFlows] list to only include hikes that have not yet been loaded from OSM.
+   * Retrieves those hikes' bounding boxes and way points from the OSM repository.
+   *
+   * The bounding box and way points are then stored in the respective hikes, the state flows are
+   * updated directly, no need for a return value.
+   *
+   * This is a one-time operation, the OSM data will only be retrieved for the hikes that are in the
+   * list at the time of calling this function.
+   *
+   * @param onSuccess To be called once the data has been retrieved successfully. Do not use this
+   *   callback to retrieve the data manually. Instead, observe [hikeFlows] to be notified
+   *   automatically.
+   * @param onFailure To be called if a problem is encountered, preventing the success of the
+   *   operation.
+   */
+  fun retrieveLoadedHikesOsmData(onSuccess: () -> Unit = {}, onFailure: () -> Unit = {}) =
```

Comment:
No, it does not refresh the existing data, it will only fetch new data.

It's actually intended to retrieve OSM data for hikes that were loaded from the saved hikes repository. When you are on the saved hikes screen, you first load the saved hikes, but then you also need their waypoints and elevation to display their elevation graph.

To do this, you call `loadSavedHikes`, then `retrieveLoadedHikesOsmData`, and finally `retrieveElevationDataFor` for each hike lazily.

See this diagram, the transition between the left-most and the middle column :

![](https://private-user-images.githubusercontent.com/46901347/389301768-9369b219-e9dc-4ed0-9a1b-a53e14a47de5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI2MTE0MDIsIm5iZiI6MTczMjYxMTEwMiwicGF0aCI6Ii80NjkwMTM0Ny8zODkzMDE3NjgtOTM2OWIyMTktZTlkYy00ZWQwLTlhMWItYTUzZTE0YTQ3ZGU1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTI2VDA4NTE0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJlZGM5NmUxOTFmNjU0MmZjYzMzMDIyOTliY2FkYWVjOTMyYTAzYmJlYWI5OWYxMDU3MWI3YWU3NTFiOGZiMWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.BK_n8UMAicn-v3ZNx8iDOxTcpPW3-qQFUmpPQzLdVKI)

Reply:
I can't see the diagram, but I understood the idea! Thanks!